src/Hl7.Fhir.Base/Model/Base.TypedElement.cs

```diff
@@ -0,0 +1,177 @@
+    /// The parent node of this node, or null if this is the root node.
+    /// </summary>
+    IScopedNode? Parent { get; } 
+
+    // /// <summary>
+    // /// An indication of the location of this node within the data represented by the <c>ITypedElement</c>.
+    // /// </summary>
+    // /// <remarks>The format of the location is the dotted name of the property, including indices to make
+    // /// sure repeated occurrences of an element can be distinguished. It needs to be sufficiently precise to aid
+    // /// the user in locating issues in the data.</remarks>
+    // string Location { get; }
+}
+
+internal record ScopeInformation(IScopedNode? Parent, string Name, int? Index);
+
+
+public abstract partial class Base : IScopedNode,
+    IFhirValueProvider, IResourceTypeSupplier
+{
+    [NonSerialized]
+    private ScopeInformation? _scopeInfo;
+    
+    private ScopeInformation ScopeInfo
+    {
+        get => LazyInitializer.EnsureInitialized(ref _scopeInfo, () => BuildRoot())!;
+        set => _scopeInfo = value;
+    } 
+
+    internal ScopeInformation BuildRoot(string? rootName = null) => new(null, rootName ?? TypeName, null);
+    
+    internal Base WithScopeInfo(ScopeInformation info)
+    {
+        this.ScopeInfo = info;
+        return this;
+    }
+
+    IEnumerable<ITypedElement> IBaseElementNavigator<ITypedElement>.Children(string? name) =>
+        this.GetElementPairs()
+            .Where(ep => (name == null || name == ep.Key))
+            .SelectMany<KeyValuePair<string, object>, Base>(ep => 
+                (ep.Key, ep.Value) switch
+                {
+                    (_, Base b) => (IEnumerable<Base>)[b.WithScopeInfo(new ScopeInformation(this, ep.Key, null))],
+                    (_, IEnumerable<Base> list) => list.Select((item, idx) => item.WithScopeInfo(new ScopeInformation(this, ep.Key, idx))),
+                    ("url", string s) when this is Extension => [new FhirUri(s).WithScopeInfo(new ScopeInformation(this, ep.Key, null))],
+                    ("id", string s) when this is Element => [new FhirString(s).WithScopeInfo(new ScopeInformation(this, ep.Key, null))],
+                    ("value", _) => [],
+                    _ => throw new InvalidOperationException("Unexpected system primitive in child list")
+                }
+            );
+        
+    IScopedNode? IScopedNode.Parent => ScopeInfo.Parent;
+
+    string IBaseElementNavigator<ITypedElement>.Name => ScopeInfo.Name;
+
+    // TODO:
+    //   Als wij een BackboneElement zijn, dan is onze naam niet this.TypeName maar "BackboneElement" of
+    //   "Element", afhankelijk van waar hij in de .net inheritance hierarchie zit.
+    //   HEt moet "code" zijn als dit een "Code<T>" is. Dat zijn geloof ik de afwijkingen.
+    //   Wellioht is er ook nog iets met de directe properties "Extension.url" en "Element.id" die van een
+    //   system type zijn ipv een FHIR type.
+    string? IBaseElementNavigator<ITypedElement>.InstanceType => 
```

Comment:
I think we can implement InstanceType based on the available (reflection) type information in .NET, to avoid referencing ModelInspector (that's a hack,really).

Reply:
I would love to know how, let's discuss.

Reply:
In the comments above the property I gave a hint: it's basically base.TypeName with some exceptions for `Code<T>` and backbone elements.  You can also look at the IStructureDefinitionSummary implementation on ClassMapping, since that actually bridges between .NET types and the ITYpedElement definition stuff.